### PR TITLE
Datasource: Change query filtering

### LINKED ISSUE #80531: Datasources: Confusion around query filtering
**Why is this needed**:

The Plugins Platform frequently gets questions related to query filtering in data source plugins. Over the years, it has caused a lot of confusion so it's time to reach to a conclusion on how to deal with filtering once and for all. 

The way I see it, there's a few problems related to query filtering that is causing confusion: 

1. Every query editor row has a `Disable query` button. This button maps to the [`hide`](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/common/common.gen.ts#L50) property which is defined in the base interface `DataQuery`. Unfortunately, `disable` and `hide` are not the same things. This inconsistency causes confusion on whether a certain query should be filtered out before it's executed, if it should be executed but its response should be filtered out etc etc.   
2. Currently, [`filterQuery`](https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L282) is defined as a virtual method in the DataSourceWithBackend class. A data source plugin author can optionally implement this method, and that is useful in case they want to apply some custom filtering. However, in most cases they likely just want to filter out queries that are disabled (hidden) in the UI. This is repetitive work, and it also causes UI inconsistencies in case the plugin author forgets to implement this method.  
3. Even though the `filterQuery` method is currently defined in the DataSourceWithBackend class, there's actually nothing "backend specific" about the filtering per se. The `Disable query` button is displayed even though the data source doesn't extend DataSourceWithBackend. It would be good if the `filterQuery` method could be defined somewhere that allows it to be implemented and called automatically also for those data sources that don't extend DataSourceWithBackend.

**What would you like to be added**:

Some proposals:
1. Add a new `disable` prop in the DataQuery interface. Mark `hide` as deprecated and remove it completely in Grafana 12 or similar (Grafana 11 will be too early). I think most data sources filter out queries that are hidden (disabled) before they're being executed, so therefore the word `disable` makes more sense. Note that to prevent dashboards from breaking, we'd still need to update the `hide` prop whenever the button in the query editor row is changed. 
2. If a certain query is `Disabled` in the UI, Grafana should filter it out before it's being passed to the backend. This should be the case **only** if the data source doesn't implement the `filterQuery` method. 
One could argue this is not a breaking change since this is the "expected behaviour", and we're only implementing a missing feature. Anyhow, this will for sure break a lot of dashboards out there, so a change like this needs to be behind a feature toggle and announced well in time before it's introduced. 
3. Move the `filterQuery` method to a new class that sits between `DataSourceWithBackend` and the DataSourceAPI. Datasources that don't extend DataSoureWithBackend could extend this new class and thereby implement the `filterQuery` method. This will make data source development more consistent. 

**Who is this feature for?**

Plugin authors 

**Related topics**
https://github.com/grafana/grafana/pull/79902
https://github.com/grafana/grafana/pull/74147
https://github.com/grafana/grafana/issues/47876

### PR DESCRIPTION
**Pull request by sunker**
**What is this feature?**
This PR changes query filtering in data source plugins. There are two noticeable changes: 

#### Responses associated with hidden queries will be removed automatically by Grafana
Starting from Grafana 11, responses that are associated with queries where `hide: true` will be removed by Grafana. Queries where `hide: true` will still be executed (unless they're prevented from it in the data source `filterQuery` method), but the associated responses will no longer be passed to the panel. To align the UI with this behaviour (and with the DataQuery [model](https://github.com/grafana/grafana/blob/187114ddd30d6a1a81e4dea1d3ab1b342a4307fb/packages/grafana-schema/src/common/common.gen.ts#L50)), the tooltip in the query editor row is changed from `Disable query` to `Hide/Show response`.  

#### The `datasource.filterQuery` method will be called from the query runner

Data source plugins that extended `DataSourceWithBackend` could optionally implement `filterQuery` to prevent certain queries from being executed. The filterQuery method was called fromwithin the [`query`](https://github.com/grafana/grafana/blob/187114ddd30d6a1a81e4dea1d3ab1b342a4307fb/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L129) method in the `DataSourceWithBackend` class. Not all data source plugins extends `DataSourceWithBackend`, and this led to inconsistencies in how filtering worked between different types of data source plugins. 

From Grafana 11, the `filterQuery` method is called from the [query runner](https://github.com/grafana/grafana/blob/187114ddd30d6a1a81e4dea1d3ab1b342a4307fb/public/app/features/query/state/runRequest.ts#L206) before the data source query method is called. This means query filtering now works in a consistent way for all types of data source plugins. 

For data source plugins that extend `DataSourceWithBackend`, it means they can no longer test query filtering by calling the `query` method, as filtering occurs at an earlier stage in the query execution flow.   

**Why do we need this feature?**

To harmonize query execution flow in backend- and non backend data source plugins and clarify the purpose of the `Hide/Show response` button for plugin authors and end-users.

**Who is this feature for?**

Plugin authors and end-users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/80531

# Release notice breaking change

For data sources that extend `DataSourceWithBackend`, the `filterQuery` method is now called **before** the data source `query` method. If the `filterQuery` method assumes that some kind of query migration happens before this method is called, you now need to do the migration inside this method. 

Users of data source plugins that did not previously remove hidden queries will see a change of behaviour: Before this change, clicking the `Disable query` button had no impact on the query result, but starting from Grafana 11 responses associated with hidden queries will no longer be returned to the panel.

### CODE REVIEW

**Review by aangelisc — APPROVED** [review]
LGTM, nice work 😊

**Comment by joshhunt** [comment]
nice 👏  It was baffling why this was left up to individual data sources in the first place! 🚀

File: public/app/features/query/state/runRequest.ts [hunk 1]
```diff
@@ -195,6 +201,15 @@ export function callQueryMethod(
       : t
   );
 
+  // do not filter queries in case a custom query function is provided (for example in variable queries)
+  if (!queryFunction) {
+    request.targets = request.targets.filter((t) => datasource.filterQuery?.(t) ?? true);
```
Comment (andresmgot) [inline public/app/features/query/state/runRequest.ts:206]: this confuses me a bit. Doesn't this mean that the target is filtered _before_ being executed?

Comment (andresmgot) [reply public/app/features/query/state/runRequest.ts:206]: sorry, I got confused with hidden / filtered :upside_down_face:

**Review by andresmgot — APPROVED** [review]
LGTM!

**Review by mckn — APPROVED** [review]
LGTM!

**Review by bohandley — APPROVED** [review]
Everything looks ok from Graphite! I can hide a query(#A) that is referenced by another query(#B) and #B is evaluated and visualized as it should. 🚀

**Review by mckn — APPROVED** [review]
LGTM!

**Review by ifrost — APPROVED** [review]

**Pull request merged**